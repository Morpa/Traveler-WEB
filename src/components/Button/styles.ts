import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'

import { ButtonProps } from '.'

export type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'color' | 'textColor'>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,

  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,

  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.large};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,

  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, hasIcon, color, textColor }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors[textColor!]};
    border: 0;
    cursor: pointer;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    text-decoration: none;
    background: ${theme.colors[color!]};
    transition: backgroud, color, ${theme.transition.default};

    &:hover {
      background: ${darken(0.1, theme.colors[color!])};
    }

    ${!!size && wrapperModifiers[size](theme)};
    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
  `}
`
