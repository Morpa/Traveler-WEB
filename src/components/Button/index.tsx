import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  color?: 'orange' | 'blueLow' | 'green'
  textColor?: 'primary' | 'textColor'
  icon?: React.ReactNode
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  {
    children,
    icon,
    size = 'medium',
    color = 'blueLow',
    textColor = 'textColor',
    ...props
  },
  ref
) => (
  <S.Wrapper
    size={size}
    ref={ref}
    hasIcon={!!icon}
    color={color}
    textColor={textColor}
    {...props}
  >
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default forwardRef(Button)
