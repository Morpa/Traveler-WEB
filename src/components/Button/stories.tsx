import { Story, Meta } from '@storybook/react/types-6-0'
import { Whatsapp } from '@styled-icons/boxicons-logos/Whatsapp'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'Acesso Restrito'
}

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />

withIcon.args = {
  size: 'medium',
  children: 'Entre em contacto',
  color: 'green',
  icon: <Whatsapp />,
  textColor: 'primary'
}

export const asLink: Story<ButtonProps> = (args) => <Button {...args} />
asLink.args = {
  size: 'large',
  children: 'Descobrir todos os lugares',
  as: 'a',
  href: 'link',
  color: 'orange',
  textColor: 'primary'
}
