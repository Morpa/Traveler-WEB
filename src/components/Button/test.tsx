import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { Whatsapp } from '@styled-icons/boxicons-logos/Whatsapp'

import Button from '.'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Fale conosco</Button>)

    expect(screen.getByRole('button', { name: /Fale conosco/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Fale conosco</Button>)

    expect(screen.getByRole('button', { name: /Fale conosco/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Fale conosco</Button>)

    expect(screen.getByRole('button', { name: /Fale conosco/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.8rem'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<Whatsapp data-testid="icon" />}>Fale conosco</Button>
    )

    expect(screen.getByText(/Fale conosco/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Fale conosco
      </Button>
    )

    expect(screen.getByRole('link', { name: /Fale conosco/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
