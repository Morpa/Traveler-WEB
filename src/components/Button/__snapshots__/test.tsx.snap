// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #617480;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #DDE9F0;
  -webkit-transition: backgroud,color,0.3s ease-in-out;
  transition: backgroud,color,0.3s ease-in-out;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: #bad2e0;
}

<button
  class="c0"
  color="blueLow"
>
  <span>
    Fale conosco
  </span>
</button>
`;
